src/app/service.rs

```diff
@@ -78,49 +83,39 @@ impl<T: Send + Sync> Service<http::Request<Incoming>> for AppService<T> {
             )
         };
 
-        // Preallocate for the middleware stack.
-        //
-        // In the future, we can cache lazily resolved middleware stacks to
-        // avoid this allocation and limit the atomic operations that occur
-        // during route resolution to 1 per dynamic param name.
-        let mut next = Next::new(VecDeque::with_capacity(8));
+        // Get a reference to the component parts of the request as well as a
+        // mutable reference to the path parameters.
+        let RequestHead {
+            ref parts,
+            ref mut params,
+            ..
+        } = *request.head_mut();
 
-        // Get a mutable reference to the component parts of the request as
-        // well as the vec that contains the path parameters.
-        let RequestHead { parts, params, .. } = request.head_mut();
-
-        // 1 atomic op per matched middleware fn and an additional op if the
-        // path segment matched a dynamic segment.
-        //
-        // Allocations only occur if a path segment has 2 :dynamic or *wildcard
-        // patterns and a static a static pattern that matches the path
-        // segment.
+        // Populate the middleware stack with the resolved routes.
         for (route, param) in self.app.router.traverse(parts.uri.path()) {
-            // Add the matched route's middleware to the call stack.
-            next.extend(route.map(Arc::clone));
+            // Extend the deque with the matching route's middleware.
+            deque.extend(route.cloned());
 
             if let Some((name, range)) = param {
-                // Include the resolved dynamic parameter in params.
+                // Include the route's dynamic parameter in params.
                 params.push(Arc::clone(name), range);
             }
         }
 
-        // Call the middleware stack to get a response.
-        ServeRequest(next.call(request))
+        ServeRequest {
+            // Call the middleware stack to get a response.
+            future: Next::new(deque).call(request),
```

Comment:
The scope of what can be done with the middleware stack is reduced to `Next::call`. Exactly what we want.